lidar_euclidean_cluster_detect/nodes/lidar_euclidean_cluster_detect/lidar_euclidean_cluster_detect.cpp

```diff
@@ -224,33 +224,58 @@ void publishCloudClusters(const ros::Publisher *in_publisher, const autoware_msg
-    for (auto i = in_clusters.clusters.begin(); i != in_clusters.clusters.end(); i++)
+    for (const auto & cluster : in_clusters.clusters)
     {
       autoware_msgs::CloudCluster cluster_transformed;
       cluster_transformed.header = in_header;
       try
       {
         _transform_listener->lookupTransform(in_target_frame, _velodyne_header.frame_id, ros::Time(),
                                              *_transform);
-        pcl_ros::transformPointCloud(in_target_frame, *_transform, i->cloud, cluster_transformed.cloud);
-        _transform_listener->transformPoint(in_target_frame, ros::Time(), i->min_point, in_header.frame_id,
+        pcl_ros::transformPointCloud(in_target_frame, *_transform, cluster.cloud, cluster_transformed.cloud);
+        _transform_listener->transformPoint(in_target_frame, ros::Time(), cluster.min_point, in_header.frame_id,
                                             cluster_transformed.min_point);
-        _transform_listener->transformPoint(in_target_frame, ros::Time(), i->max_point, in_header.frame_id,
+        _transform_listener->transformPoint(in_target_frame, ros::Time(), cluster.max_point, in_header.frame_id,
                                             cluster_transformed.max_point);
-        _transform_listener->transformPoint(in_target_frame, ros::Time(), i->avg_point, in_header.frame_id,
+        _transform_listener->transformPoint(in_target_frame, ros::Time(), cluster.avg_point, in_header.frame_id,
                                             cluster_transformed.avg_point);
-        _transform_listener->transformPoint(in_target_frame, ros::Time(), i->centroid_point, in_header.frame_id,
+        _transform_listener->transformPoint(in_target_frame, ros::Time(), cluster.centroid_point, in_header.frame_id,
                                             cluster_transformed.centroid_point);
 
-        cluster_transformed.dimensions = i->dimensions;
-        cluster_transformed.eigen_values = i->eigen_values;
-        cluster_transformed.eigen_vectors = i->eigen_vectors;
+        // Convex_hull
+        geometry_msgs::PolygonStamped polygon_tfed;
+        for (geometry_msgs::Point32 point : cluster.convex_hull.polygon.points) {
+            geometry_msgs::PointStamped new_point_stamped;
+            geometry_msgs::PointStamped new_point_stamped_tfed;
+            geometry_msgs::Point32 new_point32;
+            new_point_stamped.header = in_header;
+            new_point_stamped.point.x = point.x;
+            new_point_stamped.point.y = point.y;
+            new_point_stamped.point.z = point.z;
+            _transform_listener->transformPoint(in_target_frame, ros::Time(), new_point_stamped, in_header.frame_id, new_point_stamped_tfed);
+            new_point32.x = new_point_stamped_tfed.point.x;
+            new_point32.y = new_point_stamped_tfed.point.y;
+            new_point32.z = new_point_stamped_tfed.point.z;
+            polygon_tfed.polygon.points.push_back(new_point32);
```

Comment:
This would be significantly faster and avoid a couple of copies if you pre-allocated `polygon_tfed.polygon.points` with

```
polygon_tfed.polygon.points.resize(cluster.convex_hull.polygon.points.size())
```

And then referenced the properties of each `point` by ID or an iterator.

Additionally, allocating `new-point_stamped` and `new_point_stamped_tfed` outside this loop and re-using them would save ref/de-ref time.

Reply:
Thanks for that ! I finally decided to directly resize `cluster_transformed.convex_hull.polygon.points` with
```c++
cluster_transformed.convex_hull.polygon.points.resize(cluster.convex_hull.polygon.points.size());
```

It removes the need for an intermediate polygon.

`new_point_stamped` & `new_point_stamped_tfed` are now allocated outside the for loop.

Reply:
What do you think ?

Reply:
LGTM.